src/broker/groups.c

```diff
@@ -0,0 +1,750 @@
+    return;
+error:
+    if (flux_respond_error (h, msg, errno, errmsg) < 0)
+        flux_log_error (h, "error responding to groups.leave request");
+}
+
+/* A client wishes to LEAVE a group.
+ * Drop cached JOIN request and add LEAVE update to batch.
+ * N.B. response is sent before batch updates are applied.
+ */
+static void leave_request_cb (flux_t *h,
+                              flux_msg_handler_t *mh,
+                              const flux_msg_t *msg,
+                              void *arg)
+{
+    struct groups *g = arg;
+    const char *name;
+    struct group *group;
+    char errbuf[256];
+    const char *errmsg = NULL;
+
+    if (flux_request_unpack (msg, NULL, "{s:s}", "name", &name) < 0)
+        goto error;
+    if (!(group = group_lookup (g, name, false))
+        || !group->join_request) {
+        snprintf (errbuf,
+                  sizeof (errbuf),
+                  "rank %lu is not a member of %s",
+                  (unsigned long)g->ctx->rank,
+                  name);
+        errmsg = errbuf;
+        errno = ENOENT;
+        goto error;
+    }
+    if (groups_leave (g, name) < 0)
+        goto error;
+    flux_msg_decref (group->join_request);
+    group->join_request = NULL;
+    if (flux_respond (h, msg, NULL) < 0)
+        flux_log_error (h, "error responding to groups.join request");
+    return;
+error:
+    if (flux_respond_error (h, msg, errno, errmsg) < 0)
+        flux_log_error (h, "error responding to groups.join request");
+}
+
+/* Respond to a request for group membership.
+ * Return 0 on success, -1 on error with errno set.
+ * A failure to respond is just logged, not treated as an error.
+ */
+static int get_respond_one (struct groups *g,
+                            struct group *group,
+                            const flux_msg_t *msg)
+{
+    char *s;
+
+    if (!(s = idset_encode (group->members, IDSET_FLAG_RANGE)))
+        return -1;
+    if (flux_respond_pack (g->ctx->h, msg, "{s:s}", "members", s) < 0)
+        flux_log_error (g->ctx->h, "error responding to groups.get request");
+    free (s);
+    return 0;
+}
+
+/* 'group' membership has changed, respond to all pending groups.get
+ * requests.
+ */
+static void get_respond_all (struct groups *g, struct group *group)
+{
+    const flux_msg_t *request;
+
+    request = flux_msglist_first (group->watchers);
+    while (request) {
+        if (get_respond_one (g, group, request) < 0) {
+            flux_log_error (g->ctx->h,
+                            "error constructing groups.get response");
+        }
+        request = flux_msglist_next (group->watchers);
+    }
+}
+
+/* Process a groups.get request for group membership.  To avoid bugs arising
+ * from users making the request with nodeid=FLUX_NODEID_ANY on rank > 0,
+ * which would return that broker's subtree membership, reject such requests.
+ * If the group doesn't exist, it is created, empty.  This request may
```

Comment:
Minor question: I'm curious about the decision to create groups that do not exist with a `get`. Since groups are never destroyed, does this have the potential to accumulate unneeded group objects in the broker from accidental or `gets` with typo'd group names?

Also does this mean a client can't tell the difference between a group that doesn't exist and one that is empty (or is there no difference between the two right now?)

To conserve resources at least, would it make sense to destroy empty groups on `get` client disconnect? (Maybe that is already done and I just haven't gotten that far)

Reply:
The reason for creating a group on get and not distinguishing between empty and nonexistent was so that a streaming get request could be attached to a group before its creation.  That's also the reason why it's not removed when empty or when the joiner disconnects - there might still b a get waiting on it.

Those issues could be addressed by moving the get requests to a common list in the main groups struct.  I thought maybe it was not a huge deal given that groups are for instance owner only?  However, I may have landed in that position due to "design fatigue"!   So what do you think, should it change?

Reply:
No need to change. I was just curious.

> That's also the reason why it's not removed when empty or when the joiner disconnects - there might still b a get waiting on it.

Can you check if a get is still waiting?

This could easily be fixed in the future if it even becomes a problem, so not a big deal at all. It does seem like you could easily deal with the (admittedly unlikely) `get` to an invalid group by removing groups that are empty when the last "getter" disconnects.

But no need to do anything now I think! We need to keep forward momentum

Reply:
That's true.  OK, I'm for moving forward also, as this has taken way too long already and there is more to do in the resource module to properly handle nodes that bounce up and down or become idle.  Thanks!  I'll set MWP.